tests/prover_mock/mod.rs

```diff
@@ -100,7 +116,119 @@ impl Service {
+            if root != last_root {
+                break;
+            }
+            last_root = proof.root((*identity).into()).into();
+        }
+
+        // If the final root doesn't match the post root something's broken so we error.
+        if last_root != input.post_root {
+            return ProveResponse::failure("43", "Merkle proof verification failure.");
+        }
+
+        // If we succeed in verifying, the output should be correlated with the input,
+        // so we use the input_hash as part of it.
+        ProveResponse::success([
+            "0x2".into(),
+            input_hash,
+            "0x2413396a2af3add6fbe8137cfe7657917e31a5cdab0b7d1d645bd5eeb47ba601".into(),
+            "0x1ad029539528b32ba70964ce43dbf9bba2501cdb3aaa04e4d58982e2f6c34752".into(),
+            "0x5bb975296032b135458bd49f92d5e9d363367804440d4692708de92e887cf17".into(),
+            "0x14932600f53a1ceb11d79a7bdd9688a2f8d1919176f257f132587b2b3274c41e".into(),
+            "0x13d7b19c7b67bf5d3adf2ac2d3885fd5d49435b6069c0656939cd1fb7bef9dc9".into(),
+            "0x142e14f90c49c79b4edf5f6b7acbcdb0b0f376a4311fc036f1006679bd53ca9e".into(),
+        ])
+    }
+
+    /// Reconstructs the proof with directions as required by `semaphore-rs`.
+    ///
+    /// This allows us to utilise the proof verification procedure from that
+    /// library instead of implementing our own.
+    fn reconstruct_proof_with_directions(index: usize, proof: &[U256]) -> TreeProof {
+        let proof_vec: Vec<Branch> = proof
+            .iter()
+            .enumerate()
+            .map(|(i, node)| {
+                if Self::is_left_node_at_depth(index, i) {
+                    Branch::Left((*node).into())
+                } else {
+                    Branch::Right((*node).into())
+                }
+            })
+            .collect();
+        TreeProof { 0: proof_vec }
+    }
+
+    /// Computes whether the node at a given index is a left child or right
+    /// child.
+    ///
+    /// As the underlying tree is a binary tree, the corresponding bit to the
+    /// depth will tell us the direction. A 0 bit signifies a left child, while
+    /// a 1 bit signifies the right.
+    fn is_left_node_at_depth(index: usize, depth: usize) -> bool {
+        index & (1 << depth) == 0
+    }
+
+    /// Calculates the input hash based on the `input` parameters to the prover.
+    ///
+    /// We keccak hash all input to save verification gas. Inputs are arranged
+    /// as follows:
+    /// ```
+    /// StartIndex || PreRoot || PostRoot || IdComms[0] || IdComms[1] || ... || IdComms[batchSize-1]
+    ///     32     ||   256   ||   256    ||    256     ||    256     || ... ||     256 bits
+    /// ```
+    fn calculate_identity_registration_input_hash(input: &ProofInput) -> U256 {
+        // Calculate the input hash as described by the prover.
+        let mut hashable_bytes: Vec<u8> = vec![];
+        let mut reserved_u256_bytes: [u8; size_of::<U256>()] = Default::default();
```

Comment:
nit: I'd rename to something like `buffer` to indicate that it doesn't really do anything on its own. Looking at the name, I was expecting some special reserved block in input bytes.